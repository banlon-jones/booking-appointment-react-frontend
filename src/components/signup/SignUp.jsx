import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, Form, Container } from 'react-bootstrap';
import axios from 'axios';
import classes from './SignUp.module.css';
import CircleSpinner from '../spinners/CircleSpinner';

const SignUp = () => {
  const navigate = useNavigate();
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [passwordConfirmation, setPasswordConfirmation] = useState('');
  const [errorMessage, setErrorMessage] = useState('');
  const [loading, setLoading] = useState(false);
  const [successfullSignUp, setSuccessfullSignUp] = useState(false);

  const signUpButtonContent = () => {
    if (!loading && !successfullSignUp) {
      return 'Sign Up';
    }
    if (successfullSignUp) {
      return 'Account Created!';
    }
    return <CircleSpinner />;
  };

  useEffect(() => {
    setErrorMessage('');
  }, [password, passwordConfirmation]);

  const registerUser = async (e) => {
    e.preventDefault();
    if (password !== passwordConfirmation) {
      setErrorMessage('Passwords do not match!');
      return;
    }
    setLoading(true);
    try {
      const { data } = await axios({
        url: 'http://127.0.0.1:3000/auth/register',
        method: 'POST',
        data: {
          name,
          email,
          password,
          password_confirmation: passwordConfirmation,
        },
      });
      setLoading(false);
      if (data.error) {
        setErrorMessage(data.error);
        return;
      }
      setSuccessfullSignUp(true);
      setInterval(() => {
        navigate('/');
      }, 2000);
    } catch (err) {
      setLoading(false);
      setErrorMessage(err.message);
    }
  };

  return (
    <div className={classes.register_container}>
      <Container className={classes.form_container}>
        <h3 className={classes.form_title}>Sign Up</h3>
        <span className={classes.error_message}>{errorMessage}</span>
        <Form onSubmit={(e) => registerUser(e)}>
          <Form.Group className="mb-3" controlId="formBasicName">
            <Form.Label>Name</Form.Label>
            <span>*</span>
            <Form.Control
              type="text"
              placeholder="Enter Your Name"
              onChange={(e) => setName(e.target.value)}
              required
            />
          </Form.Group>
          <Form.Group className="mb-3" controlId="formBasicEmail">
            <Form.Label>Email address</Form.Label>
            <span>*</span>
            <Form.Control
              type="email"
              placeholder="Enter email"
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </Form.Group>
          <Form.Group className="mb-3" controlId="formBasicPassword">
            <Form.Label>Password</Form.Label>
            <span>*</span>
            <Form.Control
              type="password"
              placeholder="Password"
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </Form.Group>
          <Form.Group
            className="mb-3"
            controlId="formBasicPasswordConfirmation"
          >
            <Form.Label>Password Confirmation</Form.Label>
            <span>*</span>
            <Form.Control
              type="password"
              placeholder="Password Confirmation"
              onChange={(e) => setPasswordConfirmation(e.target.value)}
              required
            />
          </Form.Group>
          <Button
            variant="primary"
            type="submit"
            className={classes.submit_btn}
            id="sign-up-btn"
          >
            {signUpButtonContent()}
          </Button>
        </Form>
      </Container>
    </div>
  );
};

export default SignUp;
