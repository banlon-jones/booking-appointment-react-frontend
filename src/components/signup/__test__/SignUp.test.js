import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import SignUp from '../SignUp';

describe('snapshots', () => {
  const initialState = { user: { loggedIn: false, role: '' } };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  it('matches SignUp snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <SignUp />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('form fields rendering', () => {
    beforeEach(() => {
      render(
        <Provider store={store}>
          <Router>
            <SignUp />
          </Router>
        </Provider>,
      );
    });
    it("renders with 'Name' field", () => {
      const nameField = screen.getAllByLabelText('Name');
      expect(nameField[0]).toBeInTheDocument();
    });

    it("renders with 'Email' text", () => {
      const emailField = screen.getAllByLabelText('Email Address');
      expect(emailField[0]).toBeInTheDocument();
    });

    it("renders with 'Password' field", () => {
      const passwordField = screen.getAllByLabelText('Password');
      expect(passwordField[0]).toBeInTheDocument();
    });

    it("renders with 'Password Confirmation' field", () => {
      const passwordConfirmationField = screen.getAllByLabelText(
        'Password Confirmation',
      );
      expect(passwordConfirmationField[0]).toBeInTheDocument();
    });
  });
});
