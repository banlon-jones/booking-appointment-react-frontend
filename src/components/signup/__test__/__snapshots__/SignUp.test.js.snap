// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots matches SignUp snapshot 1`] = `
<div
  className="register_container"
>
  <div
    className="form_container container"
  >
    <h3
      className="form_title"
    >
      Sign Up
    </h3>
    <span
      className="error_message"
    />
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicName"
        >
          Name
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicName"
          onChange={[Function]}
          placeholder="Enter Your Name"
          required={true}
          type="text"
        />
      </div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicEmail"
        >
          Email Address
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicEmail"
          onChange={[Function]}
          placeholder="Enter email"
          required={true}
          type="email"
        />
      </div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicPassword"
        >
          Password
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicPassword"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
      </div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicPasswordConfirmation"
        >
          Password Confirmation
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicPasswordConfirmation"
          onChange={[Function]}
          placeholder="Password Confirmation"
          required={true}
          type="password"
        />
      </div>
      <button
        className="submit_btn btn btn-primary"
        disabled={false}
        id="sign-up-btn"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</div>
`;
